### PR TITLE
Update to use local-dynamo, start local-dynamo from setUp

### PR DESCRIPTION
**Pull request by chaosgame**
Hello @xiao, 

Please review the following commits I made in branch 'nathan-AttributeValueList'.

a9499a96b9105cd75ae9057b1dea023d7eba5bed (2015-07-23 17:04:42 -0700)
Update to use local-dynamo, start local-dynamo from setUp

7b973b507ef9ec53d3e86e2a2650a315b7d6b4e6 (2015-07-23 16:37:35 -0700)
Migrate all uses of AttribueValueList/ComparisonOperator to ConditionBuilder

R=@xiao

Test plan: 'New unit test, old unit tests'

### CODE REVIEW

**Comment by chaosgame** [comment]
Now that we have local-dynamo and the test code runs on it's own, it would be nice to see if we could set up travis-ci integration for the repo.

File: lib/DynamoRequest.js [hunk 1]
```diff
@@ -95,32 +80,52 @@ DynamoRequest.prototype.setReturnValues = function (returnValues) {
...
-      }
-      if (attrList) {
-        this.data.ScanFilter[key].AttributeValueList = attrList
-      }
+/**
+ * @param {Array.<FilterBuilder>} conditions An array of conditions, possibly null to indicate
+ *     no conditions.
+ * @return {DynamoRequest}
+ */
+DynamoRequest.prototype.setQueryFilter = function (conditions) {
```
Comment (xiao) [inline lib/DynamoRequest.js]: `conditions` -> `filters`?

File: lib/DynamoRequest.js [hunk 2]
```diff
@@ -95,32 +80,52 @@ DynamoRequest.prototype.setReturnValues = function (returnValues) {
...
+ * @param {Array.<FilterBuilder>} conditions An array of conditions, possibly null to indicate
+ *     no conditions.
+ * @return {DynamoRequest}
+ */
+DynamoRequest.prototype.setKeyConditions = function (conditions) {
+  this.data.KeyConditions = {}
+  if (conditions) {
+    assert.ok(Array.isArray(conditions), 'Expected array')
+    for (var i = 0; i < conditions.length; i += 1) {
+      assert.ok(conditions[i] instanceof FilterBuilder, 'Expected FilterBuilder')
```
Comment (xiao) [inline lib/DynamoRequest.js]: do we expect `ConditionBuilder` or `FilterBuilder`?

File: package.json [hunk 3]
```diff
@@ -34,6 +34,7 @@
     "closure-npc": "0.1.3",
     "nodeunit": "0.9.1",
     "nodeunitq": "0.1.1"
+    "local-dynamo": "0.1.1"
```
Comment (xiao) [inline package.json]: missing `,`

File: lib/typeUtil.js [hunk 4]
```diff
@@ -113,5 +113,6 @@ function _objectToValue(obj) {
 module.exports = {
   unpackObjectOrArray: unpackObjectOrArray,
   packObjectOrArray: packObjectOrArray,
-  valueToObject: valueToObject
+  valueToObject: valueToObject,
+  objectToValue: objectToValue
```
Comment (xiao) [inline lib/typeUtil.js]: why this change?

File: test/utils/TestUtils.js [hunk 5]
```diff
@@ -36,6 +37,22 @@ utils.getMockDatabaseClient = function () {
...
 
+utils.startLocalDynamo = function () {
+  return localDynamo.launch({
+    port: 4567
+  })
+}
+
+utils.killLocalDynamo = function (proc) {
+  var defer = Q.defer()
+  proc.on('exit', function () {
```
Comment (xiao) [inline test/utils/TestUtils.js]: this seems to be a useful util function in `local-dynamo`.

File: lib/QueryBuilder.js [hunk 6]
```diff
@@ -15,7 +16,7 @@ function QueryBuilder(options) {
   Builder.call(this, options)
 
   /** @private {!Array.<{name: string, operator: string, attributes: !Array}>} */
-  this._keyConditions = []
+  this._keyConditions = new FilterBuilder()
```
Comment (xiao) [inline lib/QueryBuilder.js]: why not add filter to `QueryBuilder` as well?

Comment (xiao) [inline lib/QueryBuilder.js]: nvm, I saw you add filter to `Builder`.

File: lib/DynamoRequest.js [hunk 7]
```diff
@@ -95,32 +80,52 @@ DynamoRequest.prototype.setReturnValues = function (returnValues) {
...
-      } else {
-        // All other ops have 1 argument.
-        attrList = [typeUtil.valueToObject(filter[key][1])]
-      }
+/**
+ * @param {Array.<FilterBuilder>} conditions An array of conditions, possibly null to indicate
+ *     no conditions.
+ * @return {DynamoRequest}
+ */
+DynamoRequest.prototype.setKeyConditions = function (conditions) {
```
Comment (xiao) [inline lib/DynamoRequest.js]: These three functions, `setKeyConditions`, `setQueryFilter`, and `setScanFilter` have most code duplicated. They could be refactorized.

**Comment by xiao** [comment]
First pass done.

A high-level question -- can we consolidate `ConditionBuilder` and `FilterBuilder` into one builder, and maybe call it `ConditionBuilder`? They are essentially all conditions and they produce the same format in DynamoDB API, for `KeyConditions`, `QueryFilter` and `ScanFilter`.

File: lib/DynamoRequest.js (hunk 2 above)
Comment (chaosgame) [inline lib/DynamoRequest.js]: `FilterBuilder` unfortunately.  `ConditionBuilder` builds a structure that can only be used with `Expects`.  Eventually, I'd like to change `Expects` to use the output of `FilterBuilder` (since it can) and remove `ConditionBuilder`, but that would change existing behavior so I'd like to save that for a future PR.

File: lib/typeUtil.js (hunk 4 above)
Comment (chaosgame) [inline lib/typeUtil.js]: Hmm.  I'll remove it from this PR.  It was in anticipation of needing to implement filtering in FakeDynamo which will have to come soon.

File: test/utils/TestUtils.js (hunk 5 above)
Comment (chaosgame) [inline test/utils/TestUtils.js]: I was hoping that you wouldn't say that -- I'll put together a PR tomorrow.

**Comment by chaosgame** [comment]
`ConditionBuilder` and `FilterBuilder` don't produce the same format in DynamoDB yet, but you are correct -- that is the plan (see my comment above about changes in behavior).

**Comment by xiao** [comment]
I still think we should consolidate them in this PR, so we don't introduce a new API (`FilterBuilder`) and change it later.

Here is a proposal that I think would work. Let me know what you think about it.

 * Keep the current two APIs in `ConditionBuilder`.
 * Add the new APIs you added into `FilterBuilder` into `ConditionBuilder`.
 * Use the new `ConditionBuilder` to generate values for all these fields (DynamoDB API), `Expected`, `KeyConditions`, `QueryFilter` and `ScanFilter`.
 * Stop using `Value` + `Exist`, and use `AttributeValueList` + `ComparisonOperator` for `Expected`. This is hidden from the user of Dynamite anyway.
 * I really like your idea to keep the APIs in `QueryBuilder` unchanged.

**Comment by chaosgame** [comment]
Alright @xiao, changes have been made.  PTAL.

**Comment by xiao** [comment]
LGTM!

Beautiful piece.
